src/Microsoft.Restier.Publishers.OData/Query/RestierQueryBuilder.cs

```diff
@@ -304,6 +310,57 @@ private void HandlePropertyAccessPathSegment(ODataPathSegment segment)
                 this.queryable = ExpressionHelpers.Select(this.queryable, selectBody);
             }
         }
+
+        // This only covers entity type cast
+        // complex type cast uses ComplexCastPathSegment and is not supported by EF now
+        private void HandleCastPathSegment(ODataPathSegment segment)
+        {
+            var castSegment = (CastPathSegment)segment;
+            Type elementType = GetClrType(castSegment.CastType);
+            this.currentEntityType = castSegment.CastType;
+            this.currentType = elementType;
+            this.queryable = ExpressionHelpers.OfType(this.queryable, elementType);
+        }
+
+        private Type GetClrType(IEdmEntityType edmType)
+        {
+            IEdmModel edmModel = api.GetModelAsync().Result;
+
+            ClrTypeAnnotation annotation = edmModel.GetAnnotationValue<ClrTypeAnnotation>(edmType);
+            if (annotation != null)
+            {
+                return annotation.ClrType;
+            }
+
+            // In case user does not use Web Api OData conversion builder, fail back to IModelMapper
+            var name = edmType.Name;
+            var namespaceName = edmType.Namespace;
+            Type elementType = null;
+
+            var mapper = api.Context.GetApiService<IModelMapper>();
+            if (mapper != null)
+            {
+                if (namespaceName == null)
+                {
+                    mapper.TryGetRelevantType(api.Context, name, out elementType);
```

Comment:
The `name` is NOT the type name but the resource name or operation name. So model mapper cannot be called like this.